##  We will now consider the Boston housing data set, from the ISLP library.
 
 
# (a) Based on this data set, provide an estimate for the population mean of medv. Call this estimate ˆ µ

In [1]:
from ISLP import load_data
import pandas as pd
import numpy as np

#Download the data
boston_data = load_data('Boston')
medv = boston_data['medv']

mean_medv = np.mean(medv)
print("Estimate of the population mean (μ̂):", mean_medv)

Estimate of the population mean (μ̂): 22.532806324110677


# (b) Provide an estimate of the standard error of ˆ µ. Interpret this result.

In [2]:
std_error_mean = np.std(medv, ddof=1) / np.sqrt(len(medv))
print("Standard error of the mean (SE(μ̂)):", std_error_mean)

Standard error of the mean (SE(μ̂)): 0.40886114749753505


# (c) Now estimate the standard error of ˆ µ using the bootstrap. How does this compare to your answer from (b)?

In [3]:
from sklearn.utils import resample

bootstrap_means = [np.mean(resample(medv)) for _ in range(1000)]
bootstrap_std_error_mean = np.std(bootstrap_means)
print("Bootstrap estimate of the standard error of the mean:", bootstrap_std_error_mean)

Bootstrap estimate of the standard error of the mean: 0.404131170683672


The difference between two std error is very small and similar.

# (d) Based on your bootstrap estimate from (c), provide a 95% confidence interval for the mean of medv. Compare it to the results obtained by using Boston['medv'].std() and the two standard error rule (3.9).

In [5]:
#C.I. of (c)
ci_lower_1 = mean_medv - 2 * bootstrap_std_error_mean
ci_upper_1 = mean_medv + 2 * bootstrap_std_error_mean
print("95% confidence interval for the mean_1:", (ci_lower_1, ci_upper_1))

#C.I. of (b)
ci_lower_2 = mean_medv - 2 * std_error_mean
ci_upper_2 = mean_medv + 2 * std_error_mean
print("95% confidence interval for the mean_2:", (ci_lower_2, ci_upper_2))


95% confidence interval for the mean_1: (21.724543982743334, 23.34106866547802)
95% confidence interval for the mean_2: (21.715084029115605, 23.35052861910575)


Both coincidence interval are exactly same.They are give us more reliable estimate of std error for mean.

# (e) Based on this data set, provide an estimate, ˆ µmed, for the median value of medv in the population.

In [6]:
median_medv = np.median(medv)
print("Estimate of the population median (μ̂_med):", median_medv)

Estimate of the population median (μ̂_med): 21.2


# (f) We now would like to estimate the standard error of ˆ µmed. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [7]:
bootstrap_medians = [np.median(resample(medv)) for _ in range(1000)]
bootstrap_std_error_median = np.std(bootstrap_medians)
print("Bootstrap estimate of the standard error of the median:", bootstrap_std_error_median)

Bootstrap estimate of the standard error of the median: 0.3776990733374915


The standard error od median is very small.It means that median is a stable estimate of population median.

# (g) Based on this data set, provide an estimate for the tenth percentile of medv in Boston census tracts. Call this quantity ˆ µ0.1

In [8]:
percentile_10 = np.percentile(medv, 10)
print("Estimate of the 10th percentile (μ̂_0.1):", percentile_10)

Estimate of the 10th percentile (μ̂_0.1): 12.75


# (h) Use the bootstrap to estimate the standard error of ˆ µ0.1. Comment on your findings

In [9]:
bootstrap_percentiles_10 = [np.percentile(resample(medv), 10) for _ in range(1000)]
bootstrap_std_error_percentile_10 = np.std(bootstrap_percentiles_10)
print("Bootstrap estimate of the standard error of the 10th percentile:", bootstrap_std_error_percentile_10)

Bootstrap estimate of the standard error of the 10th percentile: 0.48855386345826796


Althought the std error is relatively bigger than the throretical one(0.38),means that there has small alteration on 10th percentile data,but it still a stable estimate of 10th percentile data of population.